File: app/src/main/java/com/dm/bomber/services/JsonService.java
package com.dm.bomber.services;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

public abstract class JsonService extends Service {

    protected String url;
    protected String method;

    protected Request.Builder request;

    public JsonService(String url, String method, int... countryCodes) {
        super(countryCodes);

        this.url = url;
        this.method = method;
    }

    public JsonService(String url, int... countryCodes) {
        this(url, "POST", countryCodes);
    }

    public void run(OkHttpClient client, Callback callback, Phone phone) {
        request = new Request.Builder();

        RequestBody body = RequestBody.create(
                buildJson(phone), MediaType.parse("application/json"));

        request.url(url);
        request.method(method, body);

        client.newCall(request.build()).enqueue(callback);
    }

    public abstract String buildJson(Phone phone);
}

File: app/src/main/java/com/dm/bomber/services/core/JsonService.kt
package com.dm.bomber.services.core

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody

abstract class JsonService(
    protected val url: String,
    protected val method: String,
    vararg countryCodes: Int
) : Service(*countryCodes) {

    protected lateinit var request: Request.Builder

    constructor(url: String, vararg countryCodes: Int) : this(url, "POST", *countryCodes)

    override fun run(client: OkHttpClient, callback: Callback, phone: Phone) {
        request = Request.Builder()

        val body: RequestBody = buildJson(phone)!!.toRequestBody("application/json".toMediaType())

        request.url(url)
        request.method(method, body)

        client.newCall(request.build()).enqueue(callback)
    }

    abstract fun buildJson(phone: Phone): String?
}
